### PR TITLE
[GSoC] Refactored the `Series` and `MIMOSeries` to support `StateSpace` connection.

### PR DESCRIPTION
**Pull request by abhiphile**
Added support for `StateSpace` in `Series` and `MIMOSeries` connection.
#### Brief description of what is fixed or changed
 - `TransferFunction` & `TransferFunctionMatrix` can be used in series with `StateSpace`.
 -  For the connection having `StateSpace` in it, `.doit()` will return the equivalent `StateSpace` object.

#### Release Notes

* physics.control
  * Added support to use `StateSpace` with `Series` and `MIMOSeries`.

### CODE REVIEW

File: sympy/physics/control/tests/test_lti.py [hunk 1]
```diff
@@ -1748,3 +1748,135 @@ def test_StateSpace_functions():
     assert ss3.input_matrix == Matrix([[0, 0], [1, 0], [0, 1], [0, 0]])
     assert ss3.output_matrix == Matrix([[0, 1, 0, 0], [0, 0, 1, 0]])
     assert ss3.feedforward_matrix == Matrix([[0, 0], [0, 1]])
+
+
+def test_StateSpace_series():
+    # For SISO Systems
+    A1 = Matrix([[0, 1], [1, 0]])
```
Comment (faze-geek) [inline sympy/physics/control/tests/test_lti.py:1755]: Hi, where are these tests from ? If you can you link any references I'll go through them.
I'd suggest it is good practice to test out `MATLAB` examples as well as then you can judge accuracy of your outputs.

Comment (abhiphile) [reply sympy/physics/control/tests/test_lti.py:1755]: Hi @faze-geek I've found a issue with the current implementation of multiplication for state space objects.
### Matlab Code
```
a1 = [0, 1; 1, 0];
b1 = [0; 1];
c1 = [0, 1];
d1 = 0;
ss1 = ss(a1, b1, c1, d1);
a2 = [1,0;0,1];
b2 = [1;0];
c2 = [1,0];
d2 = 1;
ss2 =ss(a2, b2, c2, d2);
ser1 = series(ss2, ss1);
```
Output
```
>> ss1 * ss2

ans =
 
  A = 
       x1  x2  x3  x4
   x1   0   1   0   0
   x2   1   0   1   0
   x3   0   0   1   0
   x4   0   0   0   1
 
  B = 
       u1
   x1   0
   x2   1
   x3   1
   x4   0
 
  C = 
       x1  x2  x3  x4
   y1   0   1   0   0
 
  D = 
       u1
   y1   0

>> ss2 * ss1

ans =
 
  A = 
       x1  x2  x3  x4
   x1   1   0   0   1
   x2   0   1   0   0
   x3   0   0   0   1
   x4   0   0   1   0
 
  B = 
       u1
   x1   0
   x2   0
   x3   0
   x4   1
 
  C = 
       x1  x2  x3  x4
   y1   1   0   0   1
 
  D = 
       u1
   y1   0

```
### SymPy Multiplication
```
>>> a1 = Matrix([[0, 1], [1, 0]])
>>> b1 = Matrix([[0], [1]])
>>> c1 = Matrix([[0, 1]])
>>> d1 = Matrix([[0]])
>>> a2 = Matrix([[1, 0], [0, 1]])
>>> b2 = Matrix([[1], [0]])
>>> c2 = Matrix([[1, 0]])
>>> d2 = Matrix([[1]])
>>> ss1 = StateSpace(a1, b1, c1, d1)
>>> ss2 = StateSpace(a2, b2, c2, d2)
>>> ss1*ss2
StateSpace(
Matrix([
[1, 0, 0, 0],
[0, 1, 0, 0],
[0, 0, 0, 1],
[1, 0, 1, 0]]),

Matrix([
[1],
[0],
[0],
[1]]),

Matrix([[0, 0, 0, 1]]),

Matrix([[0]]))
>>> ss2*ss1
StateSpace(
Matrix([
[0, 1, 0, 0],
[1, 0, 0, 0],
[0, 1, 1, 0],
[0, 0, 0, 1]]),

Matrix([
[0],
[1],
[0],
[0]]),

Matrix([[0, 1, 1, 0]]),

Matrix([[0]]))
```

Comment (abhiphile) [reply sympy/physics/control/tests/test_lti.py:1755]: I think there is some bug present in multiplication of the `StateSpace` which is causing problem.

Comment (faze-geek) [reply sympy/physics/control/tests/test_lti.py:1755]: You would be aware that `StateSpaceModel` is not a unique right ? Meaning one input output system can have infinite `StateSpace` representations unlike `TransferFunctionModel`. Similarly the output here might be different but not necessarily incorrect.

Comment (faze-geek) [reply sympy/physics/control/tests/test_lti.py:1755]: Consider this in python-control - 

```
import numpy as np
import control

a1 = np.array([[0, 1], [1, 0]])
b1 = np.array([[0], [1]])
c1 = np.array([[0, 1]])
d1 = np.array([[0]])

ss1 = control.ss(a1, b1, c1, d1)

a2 = np.array([[1, 0], [0, 1]])
b2 = np.array([[1], [0]])
c2 = np.array([[1, 0]])
d2 = np.array([[1]])

ss2 = control.ss(a2, b2, c2, d2)
```

Comment (faze-geek) [reply sympy/physics/control/tests/test_lti.py:1755]: Output -
```
print(ss1*ss2)

A = [[1. 0. 0. 0.]
     [0. 1. 0. 0.]
     [0. 0. 0. 1.]
     [1. 0. 1. 0.]]

B = [[1.]
     [0.]
     [0.]
     [1.]]

C = [[0. 0. 0. 1.]]

D = [[0.]]

```
```
print(ss2*ss1)

A = [[0. 1. 0. 0.]
     [1. 0. 0. 0.]
     [0. 1. 1. 0.]
     [0. 0. 0. 1.]]

B = [[0.]
     [1.]
     [0.]
     [0.]]

C = [[0. 1. 1. 0.]]

D = [[0.]]
```

Comment (faze-geek) [reply sympy/physics/control/tests/test_lti.py:1755]: This matches with the SymPy results. I'm not sure what kind of manipulations MATLAB does but I've observed it always applies minimal realization (https://github.com/sympy/sympy/issues/25340) on it's systems which is not followed in SymPy, so we can have different outputs in some instances.

File: sympy/physics/control/lti.py [hunk 2]
```diff
@@ -1673,15 +1796,26 @@ class MIMOSeries(MIMOLinearTimeInvariant):
...
-        cls._check_args(args)
-
-        if _mat_mul_compatible(*args):
+        if args and any(isinstance(arg, StateSpace) or (hasattr(arg, 'is_StateSpace_object')
+                                            and arg.is_StateSpace_object) for arg in args):
+            # Check compatibility
+            for i in range(1, len(args)):
+                if args[i].num_inputs != args[i - 1].num_outputs:
+                    raise ValueError(filldedent("""Systems with incompatible inputs and outputs "
+                        cannot be connected in MIMOSeries"""))
```
Comment (faze-geek) [inline sympy/physics/control/lti.py:1805]:
```suggestion
                if args[i].num_inputs != args[i - 1].num_outputs:
                    raise ValueError(filldedent("""Systems with incompatible inputs and outputs "
                        cannot be connected in MIMOSeries."""))
```

**Comment by abhiphile** [comment]
I've not found any examples related to the interconnection in series , but I've followed this document https://assets.press.princeton.edu/chapters/s9012.pdf. 

I've tested it by using python control and it is giving similar results which I have included in my tests.
## For SISO
```
import numpy as np
import control

a1 = np.array([[0, 1], [1, 0]])
b1 = np.array([[0], [1]])
c1 = np.array([[0, 1]])
d1 = np.array([[0]])
ss1 = control.ss(a1, b1, c1, d1)
a2 = np.array([[1, 0], [0, 1]])
b2 = np.array([[1], [0]])
c2 = np.array([[1, 0]])
d2 = np.array([[1]])
ser1 = control.series(ss1, ss2)
print(ser1)
```
**Output**
```
A = [[0. 1. 0. 0.]
     [1. 0. 0. 0.]
     [0. 1. 1. 0.]
     [0. 0. 0. 1.]]
B = [[0.]
     [1.]
     [0.]
     [0.]]
C = [[0. 1. 1. 0.]]
D = [[0.]]
```
Number of inputs and outputs
```
print(ser.ninputs)
1
print(ser.noutputs)
1
```
An another test having an interconnection with Transfer function.

```
tf1 = control.TransferFunction(np.array([1, 0]), np.array([1, 1]))
ser_tf = control.series(tf1, ss1)
print(ser_tf)
```
**Output**
```
Inputs (1): ['u[0]']
Outputs (1): ['y[0]']

A = [[-1.  0.  0.]
     [ 0.  0.  1.]
     [-1.  1.  0.]]
B = [[1.]
     [0.]
     [1.]]
C = [[0. 0. 1.]]
D = [[0.]]
```

## For MIMO
```
a3 = np.array([[4, 1], [2, -3]])  
b3 = np.array([[5, 2], [-3, -3]])  
c3 = np.array([[2, -4], [0, 1]])  
d3 = np.array([[3, 2], [1, -1]])  
a4 = np.array([[-3, 4, 2], [-1, -3, 0], [2, 5, 3]])  
b4 = np.array([[1, 4], [-3, -3], [-2, 1]])  
c4 = np.array([[4, 2, -3], [1, 4, 3]])  
d4 = np.array([[-2, 4], [0, 1]])

ss3 = control.StateSpace(a3, b3, c3, d3)
ss4 = control.StateSpace(a4, b4, c4, d4)

ser4 = control.Series(ss3, ss4)
print(ser4)
```
**Output**
```
print(ser4)

Inputs (2): ['u[0]', 'u[1]']
Outputs (2): ['y[0]', 'y[1]']
States (5): ['x[0]', 'x[1]', 'x[2]', 'x[3]', 'x[4]']
A = [[ 4.  1.  0.  0.  0.]
     [ 2. -3.  0.  0.  0.]
     [ 2.  0. -3.  4.  2.]
     [-6.  9. -1. -3.  0.]
     [-4.  9.  2.  5.  3.]]
B = [[  5.   2.]
     [ -3.  -3.]
     [  7.  -2.]
     [-12.  -3.]
     [ -5.  -5.]]
C = [[-4. 12.  4.  2. -3.]
     [ 0.  1.  1.  4.  3.]]
D = [[-2. -8.]
     [ 1. -1.]]
```

**Comment by abhiphile** [comment]
## For SISO Docs 
```
A1 = np.array([[-1]])  
B1 = np.array([[1]])  
C1 = np.array([[-1]])  
D1 = np.array([1])  
A2 = np.array([[0]])  
B2 = np.array([[1]])  
C2 = np.array([[1]])  
D2 = np.array([[0]])
ss1 = control.StateSpace(A1, B1, C1, D1)  
ss2 = control.StateSpace(A2, B2, C2, D2)  
S5 = control.series(ss1, ss2)

print(s5)
```
**Output**
```
Inputs (1): ['u[0]']
Outputs (1): ['y[0]']
States (2): ['x[0]', 'x[1]']
A = [[-1.  0.]
     [-1.  0.]]
B = [[1.]
     [1.]]
C = [[0. 1.]]
D = [[0.]]
```

File: sympy/physics/control/lti.py [hunk 3]
```diff
@@ -1739,6 +1878,19 @@ def doit(self, cancel=False, **kwargs):
...
+            # Return the equivalent StateSpace model
+            res = self.args[0]
+            if not isinstance(res, StateSpace):
+                res = res.doit().rewrite(StateSpace)
+            for arg in self.args[1:]:
+                if not isinstance(arg, StateSpace):
+                    arg = arg.doit().rewrite(StateSpace)
+                else:
+                    arg = arg.doit()
+                res = arg * res
```
Comment (abhiphile) [inline sympy/physics/control/lti.py:1891]: I've noticed an problem with the previous code so I've changed it from 
`res *= arg` to `res = arg * res`.
I've refereed to this :
 
image

Source : https://assets.press.princeton.edu/chapters/s9012.pdf

**Comment by faze-geek** [comment]
> I've not found any examples related to the interconnection in series

Okay, no issues. I've gone through the code. Let's see what others have to say.

This is a good start !

**Review by faze-geek — APPROVED** [review]

**Review by Sc0rpi0n101 — APPROVED** [review]

**Pull request merged**